src/NServiceBus.Transport.SQS/InputQueuePump.cs

```diff
@@ -388,6 +336,78 @@ await MovePoisonMessageToErrorQueue(receivedMessage, messageProcessingCancellati
+
+            if (receivedMessage.MessageAttributes.TryGetValue(TransportHeaders.Headers, out var headersAttribute))
+            {
+                transportMessage = new TransportMessage
+                {
+                    Headers = JsonSerializer.Deserialize<Dictionary<string, string>>(headersAttribute.StringValue) ?? [],
+                    Body = receivedMessage.Body
+                };
+                transportMessage.Headers[Headers.MessageId] = messageId;
+                if (receivedMessage.MessageAttributes.TryGetValue(TransportHeaders.S3BodyKey, out var s3BodyKey))
+                {
+                    transportMessage.Headers[TransportHeaders.S3BodyKey] = s3BodyKey.StringValue;
+                    transportMessage.S3BodyKey = s3BodyKey.StringValue;
+                }
+            }
+            else
+            {
+                // When the MessageTypeFullName attribute is available, we're assuming native integration
+                if (receivedMessage.MessageAttributes.TryGetValue(TransportHeaders.MessageTypeFullName, out var enclosedMessageType))
+                {
+                    var headers = new Dictionary<string, string>
+                    {
+                                { Headers.MessageId, messageId },
+                                { Headers.EnclosedMessageTypes, enclosedMessageType.StringValue },
+                                {
+                                    TransportHeaders.MessageTypeFullName, enclosedMessageType.StringValue
+                                } // we're copying over the value of the native message attribute into the headers, converting this into a nsb message
+                            };
+
+                    if (receivedMessage.MessageAttributes.TryGetValue(TransportHeaders.S3BodyKey, out var s3BodyKey))
+                    {
+                        headers.Add(TransportHeaders.S3BodyKey, s3BodyKey.StringValue);
+                    }
+
+                    transportMessage = new TransportMessage
+                    {
+                        Headers = headers,
+                        S3BodyKey = s3BodyKey?.StringValue,
+                        Body = receivedMessage.Body
+                    };
+                }
+                else
+                {
+                    transportMessage = JsonSerializer.Deserialize<TransportMessage>(receivedMessage.Body, transportMessageSerializerOptions);
+                    // HINT: The message cannot be processed without headers and will throw
+                    if (transportMessage?.Headers == null)
+                    {
+                        throw new Exception("Transport message is missing headers element");
+                    }
+                    // HINT: Message Id is the only required header
+                    if (!transportMessage.Headers.ContainsKey(Headers.MessageId))
+                    {
+                        transportMessage.Headers[Headers.MessageId] = messageId;
+                    }
```

Comment:
TODO: move this check out of the `else` right before returning the `transportMessage`